src/libcore/python/xml_v.cpp

```diff
@@ -38,8 +49,226 @@ MTS_PY_EXPORT(xml) {
-            return cast_object(
-                xml::load_string(name, mitsuba::detail::get_variant<Float, Spectrum>(), param));
+            return cast_object(xml::load_string(name, GET_VARIANT(), param));
         },
         "string"_a, D(xml, load_string));
+
+    m.def(
+        "load_dict",
+        [](const py::dict dict) {
+            std::map<std::string, ref<Object>> instances;
+            return cast_object(load_dict<Float, Spectrum>(dict, instances));
+        },
+        "dict"_a,
+R"doc(Load a Mitsuba scene or object from an Python dictionary
+
+Parameter ``dict``:
+    Python dictionary containing the object description
+
+)doc");
+}
+
+// Helper function to find a value of "type" in a Python dictionary
+std::string get_type(const py::dict &dict) {
+    for (auto& [key, value] : dict)
+        if (key.cast<std::string>() == "type")
+            return value.cast<std::string>();
+
+    Throw("Missing key 'type' in dictionary: %s", dict);
+}
+
+#define SET_PROPS(PyType, Type, Setter)         \
+    if (py::isinstance<PyType>(value)) {        \
+        props.Setter(key, value.cast<Type>());  \
+        continue;                               \
+    }
+
+/// Helper function to give the object a chance to recursively expand into sub-objects
+void expand_and_set_object(Properties &props, const std::string &name, const ref<Object> &obj) {
+    std::vector<ref<Object>> children = obj->expand();
+    if (children.empty()) {
+        props.set_object(name, obj);
+    } else if (children.size() == 1) {
+        props.set_object(name, children[0]);
+    } else {
+        int ctr = 0;
+        for (auto c : children)
+            props.set_object(name + "_" + std::to_string(ctr++), c);
+    }
+}
+
+template <typename Float, typename Spectrum>
+ref<Object> load_dict(const py::dict &dict, std::map<std::string, ref<Object>> &instances) {
+
+    std::string type = get_type(dict);
+    bool is_scene = (type == "scene");
+
+    const Class *class_;
+    if (is_scene)
+        class_ = Class::for_name("Scene", GET_VARIANT());
+    else
+        class_ = PluginManager::instance()->get_plugin_class(type, GET_VARIANT());
+
+    bool within_emitter = (class_->parent()->alias() == "emitter");
+    Properties props(type);
+
+    for (auto& [k, value] : dict) {
+        std::string key = k.cast<std::string>();
+
+        if (key == "type")
+            continue;
+
+        if (key == "id") {
+            props.set_id(value.cast<std::string>());
+            continue;
+        }
+
+        SET_PROPS(py::bool_, bool, set_bool);
+        SET_PROPS(py::int_, int64_t, set_long);
+        SET_PROPS(py::float_, Properties::Float, set_float);
+        SET_PROPS(py::str, std::string, set_string);
+        SET_PROPS(Properties::Array3f, Properties::Array3f, set_array3f);
+        SET_PROPS(Properties::Transform4f, Properties::Transform4f, set_transform);
+
+        // Cast py::list and numpy.ndarray to Array3f
+        if (py::isinstance<py::list>(value) ||
```

Comment:
You can make it more general and accept any type here (even `tuple` or other classes with the right duck typing) by attempting a cast to some reasonable type (``std::array<float, 3>`` or directly ``Array3f``). Note that this may fail with a `py::cast_error`, but that is something you can catch.

Reply:
Also note the dangers of directly calling Python C API functions. For example, `PyObject_Length` can fail and return `-1` in which case, an error state is set. Continuing without resetting this state can have unintended consequences. Pybind11 wraps many python function calls and converts such errors into an exception of type `py::error_already_set` (which means that there is a C++ exception, and it's linked to a currently set Python error object). If you catch and ignore such an exception, its destructor will clear the Python error state.

Reply:
Great suggestion!